### DISCUSSION TITLE
netlink: Error: interval overlaps with an existing one

### DISCUSSION THREAD

**Discussion by dark-penguin** [FoobarOy/foomuuri#52] [General]
Hi!
Thanks for the great project; I don't know what else could have replaced Shorewall for me! :heart: 

I have two interfaces with obviously different subnets:
```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.0.101  netmask 255.255.255.0  broadcast 192.168.0.255

virbr0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.122.1  netmask 255.255.255.0  broadcast 192.168.122.255
```
I add them to separate zones in the config file, which seems to work fine:
```
zone {
    localhost
    public eth0
    libvirt virbr*
}
```
But if I add eth0 to the same zone (or any other zone) via NetworkManager, then all eth0 connectivity is lost.

I see those messages in the logs:
```
$ sudo systemctl status foomuuri-dbus.service
<...>
Oct 07 22:04:35 MSI foomuuri[2063]: Interface "eth0" change to zone "public"
Oct 07 22:04:35 MSI foomuuri[2063]: netlink: Error: interval overlaps with an existing one
Oct 07 22:04:35 MSI foomuuri[2063]: netlink: Error: interval overlaps with an existing one
```
I suppose NetworkManager is having trouble figuring out the subnet for my (DHCP-configured) eth0 interface? Or is it Foomuuri that fails to correctly determine its "interval"?

Sorry for being on Debian Bookworm... NetworkManager is pretty old, but Foomuuri is from backports, so it's not that much outdated.
```
Package: foomuuri
Version: 0.24-2~bpo12+1
Package: network-manager
Version: 1.42.4-1
```

**Comment by dark-penguin** [comment]
Actually, seems like there's more trouble with virbr0. That's the "default NAT bridge" that virt-manager has built-in. It is set to automatically start at boot.

Here is my Foomuuri configuration:
```
zone {
    localhost
    public   eth0
    libvirt  virbr*
}

localhost-any {
    accept
}

libvirt-libvirt {
    accept
}

libvirt-any {
    accept
}
```
If I configure the network on a VM manually, then it works (the VM has Internet connectivity). But the VM can not receive its network configuration via DHCP. I don't know much about the internals of this "default network", so it is possible that the "virtual DHCP server" is actually not on the host's IP, which might be surprising for Foomuuri.

Changing virbr0 subnet to something vastly different (10.8.0.0/24) did not help, so seems like it's not even about the subnet.

If I change "virbr*" to "virbr0", then the message about "overlapping intervals" is no longer there! Assigning the zone via NetworkManager works! But VMs still can't get configuration via DHCP.

**Comment by kimheino** [comment]
To your first problem:

> Oct 07 22:04:35 MSI foomuuri[2063]: netlink: Error: interval overlaps with an existing one

That confusing error message from `nft` is referring to interface name conflict.

You have defined that wildcard interface `virbr*` belongs to libvirt zone. NetworkManager/D-Bus is trying to add another (new) interface `virbr0` to some zone. That causes virbr* and virbr0 to conflicts.

If you are using NM/D-Bus then you can simply define zones without any interface:

```
zone {
    localhost
    public  
    libvirt 
}
```

Or you can define zones with interfaces, but without wildcards:

```
zone {
    localhost
    public   eth0
    libvirt  virbr0 virbr1 virbr2 virbr3 virbr4
}
```

**Comment by kimheino** [comment]
DHCP problem:

> libvirt-any {
>    accept
> }

That actually doesn't accept all traffic. Broadcast/multicast are still dropped and DHCP is using broadcast. Try adding rule `broadcast accept`to all zone-zone sections.

Update:

It's enough to add it to libvirt-any (libvirt-localhost):

```
libvirt-any {
    broadcast accept
    accept
}
```

**Comment by dark-penguin** [comment]
Thank you! Both suggestions helped.

If I understand correctly:
- `broadcast accept` should be added before `accept`, because otherwise traversing the rules will stop at `accept`
- I can add only `dhcp-client` instead if I only want DHCP
- But do I not need `dhcp-server` to also allow the DHCP server to reply with another broadcast?..

**Comment by kimheino** [comment]
> broadcast accept should be added before accept, because otherwise traversing the rules will stop at accept

Foomuuri will sort broadcast-specific rules first, so order in this specific case doesn't really matter. I still recommend writing specific rules first (broadcast accept) and final policy (accept) last.

> I can add only dhcp-client instead if I only want DHCP
> But do I not need dhcp-server to also allow the DHCP server to reply with another broadcast?..

```
libvirt-localhost {
  dhcp-server  # request packets to DHCP server on localhost
}

localhost-libvirt {
  dhcp-client  # reply packets to DHCP client on libvirt
}
```

DHCP is confusing. Shorewall's dhcp-macro included both directions. Foomuuri can't do that, you must accept both ways.

**Comment by dark-penguin** [comment]
Thank you for your help! I can now go spread the glory of Foomuuri. :slightly_smiling_face: 

Note to potential future readers, including myself: if it doesn't look like a bug, ask in [discussions](https://github.com/FoobarOy/foomuuri/discussions) instead.

**Comment by kimheino** [comment]
I converted this issue to discussion for future readers. I'll also document wildcard interfaces to wiki. They are currently undocumented feature.
